source/utils/adios_iotest/adios_remote_read.cpp

```diff
@@ -0,0 +1,672 @@
+/*
+ *  Test program for remote reading
+ * Created 9/27/2023
+ *      Author: Dmitry Ganyushin ganyushin@gmail.com
+ */
+#include <chrono>
+#include <fstream>
+#include <getopt.h>
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+#include <vector>
+
+#include <adios2.h>
+
+#if ADIOS2_USE_MPI
+#include <mpi.h>
+#endif
+bool DEBUG = false;
+enum test_cases
+{
+    DIM1,
+    DIM3X,
+    DIM3Y,
+    DIM3Z,
+    DIM3PLANEYZ,
+    DIM3PLANEXY,
+    DIM3PLANEXZ
+};
+void read1D(int nproc, int rank, const std::string &filename, const int NSTEPS, adios2::IO &io,
+            std::vector<std::string> &variables, int output_length);
+/* test 2 and 3
+ * read 1 or many 1D variables
+ */
+void read3D(int nproc, int rank, const std::string &filename, const int NSTEPS, adios2::IO &io,
+            std::vector<std::string> &variables, int direction, double ratio,
+            int output_line_length);
+/* test 5
+ * A 3D subset from 3D variable
+ */
+
+void read3DPlane(int nproc, int rank, const std::string &filename, const int NSTEPS, adios2::IO &io,
+                 std::vector<std::string> &variables, int direction, double ratio,
+                 int output_line_length);
+
+std::string &getOutputString(int rank, int output_line_length, std::string &out,
+                             std::vector<double> &data);
+
+/* test 5
+ * A 3D subset from 3D variable
+ */
+
+void read1D(int nproc, int rank, const std::string &filename, const int NSTEPS, adios2::IO &io,
+            std::vector<std::string> &variables_in, double ratio, int output_line_length)
+{
+    unsigned int startX;
+    unsigned int countX;
+
+    try
+    {
+        std::chrono::time_point<std::chrono::system_clock> start;
+        std::chrono::time_point<std::chrono::system_clock> end;
+        start = std::chrono::system_clock::now();
+
+        adios2::Engine reader = io.Open(filename, adios2::Mode::Read);
+
+        /* get variables with 1D shape */
+
+        std::string out;
+        for (size_t step = 0; step < NSTEPS; step++)
+        {
+            reader.BeginStep();
+            auto variables = io.AvailableVariables(true);
+
+            for (auto const &var_name : variables)
+            {
+                if (!variables_in.empty() && var_name.first != variables_in[0])
+                    continue;
+                adios2::Variable<double> var = io.InquireVariable<double>(var_name.first);
```

Comment:
InquireVariable will return NULL if the variable exists but isn't of type double.  You should probably check for this possibility rather than segfaulting on the next line.

Reply:
Yes, read functions should be generic, but currently supports only one type. The loop was made mainly for future development, for matching multiple variables from the input list.

Reply:
OK, but I don't think you'd ever actually do it that way with multiple variables.  Those would pretty naturally come in as a list that you iterated over and do InqVar on, which would be vastly easier than iterating through all the variables and then seeing if each one was listed.  Alternatively, if you actually supported types you'd call VariableType first to see if it existed and then a templated InqVar to get the actual handle.  So, I don't think you'd ever use this loop structure...

Reply:
In the expression 
``` adios2::Variable<double> var = io.InquireVariable<double>(var_name.first);```
```var``` cannot return nullptr or NULL. Do we have a test for that? Maybe things got broken.
The compiler does not understand ```if (var != nullptr)```

Reply:
No, because var is not a pointer here so you can't compare it to null.  But the Variable class does have a boolean operator, so you can do `if (var)` instead of `if (var != nullptr)`.  That's probably the test that you want (and it does appear in many, many places in our tests).